File: src_fortran/sub/regworld.f
c
c
c
c
c
c
      subroutine REGWORLD(ityp,cord1,cord2,place,nreg)
      implicit none
      integer ityp,nreg,last,num,in,nlo,latin,mhemis,i,iuk
      integer nval,i1,i2
      real cord1,cord2,xnrlon,galat,galon,pi
      real galor,galar,gelat,gelon,xazr,xdr,q,as
      real pirad,pi2,gelac,xlatr,gelor,ck1,ck2,ac
      real alon,xlat,xlon,xreg
      integer indfe,lt50,lt25
      character irname*24300
      COMMON /REGIONcom/ INDFE(730),LT50(111),LT25(421),
     &                   irname,XNRLON(14400)
C     PROGRAM TO COMPUTE REGION NUMBER AND NAME BASED ON THE INPUT
C        ITYP <=0         LATITUDE&LONGITUDE
C        ITYP > 0         DISTANCE & AZIMUTH
C    THE REGION NAMES ARE FROM FLINN & ENGDAHL#S TABLES
      character place*32
      real BUF(40)
      character*1 cbk
      DATA GALAT/46.7706/,GALON/8.09428/
      data cbk/' '/
C
      PI=3.141593
      PIRAD=PI/180.
      PI2=PI*2.
      LAST=89
C
C     CONVERT ARRAY COORDINATES TO RADIANS
      GALOR=GALON*PIRAD
      GALAR=GALAT*PIRAD
      IF(ITYP) 850,850,860
C
  850 GELAT=CORD1
      GELON=CORD2
      NUM=1
      GO TO 212
C
  860 XAZR=CORD2*PIRAD
      XDR= CORD1*PIRAD
      NUM=1
C
      Q=SIN(GALAR)*COS(XDR)+COS(GALAR)*SIN(XDR)*COS(XAZR)
      IF(ABS(Q).GT.1.) GO TO 891
      AS=ASIN(Q)
      GO TO 890
C
  891 CONTINUE
      GELON=0.
      GELAT=0.
      NUM=NUM-1
      GO TO 212
C
  890 GELAC=AS
      XLATR=GELAC
      IF(COS(GELAC))45,46,45
   45 IF(COS(GALAR)) 47,46,47
   46 GELOR=0.0
      GO TO 34
C
   47 CONTINUE
      CK1=(COS(XDR)-SIN(GALAR)*SIN(XLATR))/
     *(COS(GALAR)*COS(XLATR))
C
      CK2=(SIN(XAZR)*SIN(XDR))/COS(XLATR)
      IF(CK1.GE.0.) GO TO 37
      IF(CK2.GE.0.) GO TO 882
C
      IN=1
      GO TO 881
C
  882 IN=2
      GO TO 883
C
   37 IF(CK2.GE.0.) GO TO 885
C
      IN=3
      GO TO 881
C
  885 IN=4
C
  883 IF(ABS(CK1).GT.1.0) GO TO 887
      AC=ACOS(CK1)
      GO TO 888
C
  881 IF(ABS(CK2).GT.1.0) GO TO 887
      AS=ASIN(CK2)
C
  888 GO TO (33,32,31,36),IN
C
   33 GELOR=GALOR-PI-AS
      GO TO 34
C
   32 GELOR=GALOR+AC
      GO TO 34
C
   31 GELOR=GALOR+AS
      GO TO 34
C
   36 GELOR=GALOR+AC
      GO TO 34
C
  887 CONTINUE
      GELOR=0.0
C
C     CONVERT LONGITUDE AND LATITUDE FROM RADIANS TO DEGREES
C     AND REDUCE LONGITUDE TO A FORM LESS THAN 360 DEGREES
C
   34 CONTINUE
      ALON=ABS(GELOR)
  506 IF(ALON.LT.PI2) GO TO 116
      ALON=ALON-PI2
      GO TO 506
C
  116 GELOR=SIGN(ALON,GELOR)
      IF(ABS(GELOR).GT.PI) GELOR=GELOR-SIGN(PI2,GELOR)
C
C  LONGITUDE IN DEGREES
C
      GELON=GELOR/PIRAD
C
C     LATITUDE IS CONVERTED FROM GEOCENTRIC TO GEODETIC COORDINATES
C
      GELAT=GELAC/PIRAD
C
  212 IF(NUM.EQ.1) GO TO 5000
      place=' '
      GO TO 900
C
 5000 CONTINUE
C
C      CLEAR BUFFER
C
C     ASSIGN REGION NUMBER AND NAME
C
      place=' '
  215 XLAT=GELAT
      XLON=GELON
      IF(XLON.LT.0.) XLON=XLON+360.
C
C     DETERMINE EARTH QUADRANT
C
      IF(XLAT.LT.0.) GO TO 4
      IF(XLON.GT.180.) GO TO 6
C
C     LATITUDE POSITIVE,LONGITUDE 0-180
C
      MHEMIS=0
      GO TO 9
C
C     LATITUDE POSITIVE,LONGITUDE 181-360
C
    6 MHEMIS=3600
      XLON=360.-XLON
      GO TO 9
C
    4 IF(XLON.GT.180.) GO TO 8
C
C     LATITUDE NEGATIVE,LONGITUDE 0-180
C
      MHEMIS=7200
      GO TO 9
C
C     LATITUDE NEGATIVE,LONGITUDE 181-360
C
    8 MHEMIS=10800
      XLON=360.-XLON
C
    9 CONTINUE
C
C     CHECK FOR ZERO LATITUDE
C
      IF(ABS(XLAT).GE.1.0) GO TO 13
      LATIN=0
      GO TO 17
C
   13 LATIN=ABS(XLAT)
      IF(LATIN.GT.LAST) LATIN=LAST
C
   17 CONTINUE
C
      NLO=ABS(XLON)
      I=MHEMIS + 40*LATIN + 1
      do iuk=1,40
         buf(iuk)=xnrlon(i+iuk-1)
      enddo
      NVAL=BUF(1)*1000.+1.1
      DO 56 I=2,NVAL
      IF(NLO.LT.int(BUF(I))) GOTO 57
56    CONTINUE
      I=NVAL+1
57    NREG=BUF(I-1)
      XREG=BUF(I-1)-NREG
      NREG=XREG*1000. + 0.1
      I1=INDFE(NREG)+1
      I2=INDFE(NREG+1)
      place=irname(i1:i2)
C
  900 continue
      return
C
      END ! of subr. regworld
c
cek    end of vel_topo.f
c
cek    begin of matrinv.f
c
      SUBROUTINE matrinv(N,A,B)
c
c MATRINV invertiert eine SYMMETRISCHE NxN-Matrix
c
c Aufruf: call MATRINV(N,A,B)
c
c         N = Anz. Zeilen/Kolonnen der Matrix
c         A = zu invertierende Matrix (Input)
c         B = inv(A)    (Output)
c
      DIMENSION A(N*N),B(N*N)
      B(1)= 1.0/A(1)
      IF(N-1) 60,60,2
2     NN= N*N
      DO 4 I=2,NN
4     B(I)= 0.0
      MM=1
      KN= 0
      DO 50 M=2,N
      K= M-1
      MM= MM+N+1
      KN= KN+N
      EK= A(MM)
      MI= M-N
      DO 10 I=1,K
      MI= MI+N
      IJ= I-N
      JM= KN
      DO 10 J= 1,K
      IJ= IJ+N
      JM= JM+1
10    EK= EK-A(MI)*B(IJ)*A(JM)
      B(MM)= 1.0/EK
      MI= M-N
      IM= KN
      DO 30 I=1,K
      IM= IM+1
      IJ= I-N
      JM= KN
      DO 20 J= 1,K
      IJ= IJ+N
      JM= JM+1
20    B(IM)= B(IM)-B(IJ)*A(JM)*B(MM)
      MI= MI+N
30    B(MI)= B(IM)
      IM= KN
      DO 40 I=1,K
      IM= IM+1
      MJ= M-N
      IJ= I-N
      DO 40 J=1,K
      MJ=MJ+N
      IJ= IJ+N
40    B(IJ)= B(IJ)+B(IM)*B(MJ)*EK
50    CONTINUE
60    continue
      return
      END

File: src/sub/regworld.c
#include <math.h>
#include <string.h>

extern int indfe[];
extern int lt50[];
extern int lt25[];
extern char irname[];
extern double xnrlon[];

void regworld(int ityp, double cord1, double cord2, char *place, int *nreg) {
    const double galat = 46.7706f;
    const double galon = 8.09428f;

    double gelat = 0.0;
    double gelon = 0.0;

    if (ityp <= 0) {
        gelat = cord1;
        gelon = cord2;
    } else {
        double pi = 3.141593f;
        double pirad = pi / 180.0;
        double galor = galon * pirad;
        double galar = galat * pirad;
        double xazr = cord2 * pirad;
        double xdr = cord1 * pirad;

        double q = sinf(galar) * cosf(xdr) + cosf(galar) * sinf(xdr) * cosf(xazr);
        if (fabs(q) > 1.0f) {
            place[0] = '\0';
            return;
        }

        double gelac = asinf(q);
        double xlatr = gelac;
        double gelor = 0.0;

        if (cosf(gelac) != 0.0 && cosf(galar) != 0.0) {
            double ck1 = (cosf(xdr) - sinf(galar) * sinf(xlatr)) / (cosf(galar) * cosf(xlatr));
            double ck2 = (sinf(xazr) * sinf(xdr)) / cosf(xlatr);

            if (fabs(ck1) <= 1.0f && fabs(ck2) <= 1.0f) {
                if (ck1 < 0.0 && ck2 >= 0.0) gelor = galor + acosf(ck1);
                else if (ck1 >= 0.0 && ck2 < 0.0) gelor = galor + asinf(ck2);
                else if (ck1 >= 0.0 && ck2 >= 0.0) gelor = galor + acosf(ck1);
                else gelor = galor - pi - asinf(ck2);
            }
        }

        double alon = fabs(gelor);
        while (alon >= 2.0f * pi) {
            alon -= 2.0f * pi;
        }
        gelor = copysignf(alon, gelor);
        if (fabs(gelor) > pi) gelor -= copysignf(2.0f * pi, gelor);

        gelon = gelor / pirad;
        gelat = gelac / pirad;
    }

    memset(place, ' ', 32);
    place[32] = '\0';

    double xlat = gelat;
    double xlon = gelon;
    if (xlon < 0.0) xlon += 360.0;

    int mhemis = 0;
    if (xlat >= 0.0) {
        if (xlon > 180.0) {
            mhemis = 3600;
            xlon = 360.0 - xlon;
        }
    } else {
        if (xlon > 180.0) {
            mhemis = 10800;
            xlon = 360.0 - xlon;
        } else {
            mhemis = 7200;
        }
    }

    int latin = (fabs(xlat) < 1.0f) ? 0 : (int)fabs(xlat);
    if (latin > 89) latin = 89;

    int nlo = (int)fabs(xlon);
    int i0 = mhemis + 40 * latin;

    double buf[40];
    for (int i = 0; i < 40; ++i) buf[i] = xnrlon[i0 + i];

    int nval = (int)(buf[0] * 1000.0 + 1.1f);
    int idx = 1;
    for (; idx < nval; ++idx) {
        if (nlo < (int)buf[idx]) break;
    }

    double xreg = buf[idx - 1] - (int)buf[idx - 1];
    *nreg = (int)(xreg * 1000.0 + 0.1f);
    int i1 = indfe[*nreg];
    int i2 = indfe[*nreg + 1];
    int n = i2 - i1;
    if (n > 32) n = 32;
    memcpy(place, irname + i1, (size_t)n);
    if (n < 32) place[n] = '\0';
}

void matrinv(int n, const double *a, double *b) {
    b[0] = 1.0f / a[0];
    if (n <= 1) return;

    int nn = n * n;
    for (int i = 1; i < nn; ++i) {
        b[i] = 0.0;
    }

    int mm = 0;
    int kn = 0;
    for (int m = 2; m <= n; ++m) {
        int k = m - 1;
        mm = mm + n + 1;
        kn = kn + n;

        double ek = a[mm - 1];
        int mi = m - n;
        for (int i = 1; i <= k; ++i) {
            mi = mi + n;
            int ij = i - n;
            int jm = kn;
            for (int j = 1; j <= k; ++j) {
                ij = ij + n;
                jm = jm + 1;
                ek = ek - a[mi - 1] * b[ij - 1] * a[jm - 1];
            }
        }

        b[mm - 1] = 1.0f / ek;
        mi = m - n;
        int im = kn;
        for (int i = 1; i <= k; ++i) {
            im = im + 1;
            int ij = i - n;
            int mj = kn;
            for (int j = 1; j <= k; ++j) {
                ij = ij + n;
                mj = mj + 1;
                b[im - 1] = b[im - 1] - b[ij - 1] * a[mj - 1] * b[mm - 1];
            }
            mi = mi + n;
            b[mi - 1] = b[im - 1];
        }

        im = kn;
        for (int i = 1; i <= k; ++i) {
            im = im + 1;
            int mj = m - n;
            int ij = i - n;
            for (int j = 1; j <= k; ++j) {
                mj = mj + n;
                ij = ij + n;
                b[ij - 1] = b[ij - 1] + b[im - 1] * b[mj - 1] * ek;
            }
        }
    }
}
